<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Estimation and tests of hierarchical F-statistics</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>
<h1> Estimation and tests of hierarchical F-statistics
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]">
</h1>
<hr>
<div align="center">
<a href="../../../doc/html/packages.html"><img src="../../../doc/html/left.jpg" alt="[Up]" width="30" height="30" border="0"></a>
<a href="../../../doc/html/index.html"><img src="../../../doc/html/up.jpg" alt="[Top]" width="30" height="30" border="0"></a>
</div><h2>Documentation for package &lsquo;hierfstat&rsquo; version 0.04-10</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td width="25%"><a href="allele.count.html">allele.count</a></td>
<td>Allelic counts</td></tr>
<tr><td width="25%"><a href="allelic.richness.html">allelic.richness</a></td>
<td>Estimates allelic richness</td></tr>
<tr><td width="25%"><a href="basic.stats.html">basic.stats</a></td>
<td>Basic statistics</td></tr>
<tr><td width="25%"><a href="betai.html">betai</a></td>
<td>Estimation of beta per population</td></tr>
<tr><td width="25%"><a href="boot.ppfis.html">boot.ppfis</a></td>
<td>Performs bootstrapping over loci of population's Fis</td></tr>
<tr><td width="25%"><a href="boot.ppfst.html">boot.ppfst</a></td>
<td>Performs bootstrapping over loci of pairwise Fst</td></tr>
<tr><td width="25%"><a href="boot.vc.html">boot.vc</a></td>
<td>Bootstrap confidence intervals for variance components</td></tr>
<tr><td width="25%"><a href="cfe.dist.html">cfe.dist</a></td>
<td>Estimates Cavalli-Sforza &amp; Edwards Chord distance</td></tr>
<tr><td width="25%"><a href="da.dist.html">da.dist</a></td>
<td>Estimates Nei's DA distance</td></tr>
<tr><td width="25%"><a href="eucl.dist.html">eucl.dist</a></td>
<td>Estimates euclidian distances</td></tr>
<tr><td width="25%"><a href="eucl.dist.trait.html">eucl.dist.trait</a></td>
<td>calculates euclidian distance among populations for a trait</td></tr>
<tr><td width="25%"><a href="exhier.html">exhier</a></td>
<td>Example data set with 4 levels, one diploid and one haploid locus</td></tr>
<tr><td width="25%"><a href="g.stats.html">g.stats</a></td>
<td>Calculates likelihood-ratio G-statistic on contingency table</td></tr>
<tr><td width="25%"><a href="g.stats.glob.html">g.stats.glob</a></td>
<td>Likelihood ratio G-statistic over loci</td></tr>
<tr><td width="25%"><a href="genot2al.html">genot2al</a></td>
<td>Separates diploid genotypes in its constituant alleles</td></tr>
<tr><td width="25%"><a href="getal.html">getal</a></td>
<td>Converts diploid genotypic data into allelic data</td></tr>
<tr><td width="25%"><a href="getal.b.html">getal.b</a></td>
<td>Converts diploid genotypic data into allelic data</td></tr>
<tr><td width="25%"><a href="gtrunchier.html">gtrunchier</a></td>
<td>Genotypes at 6 microsatellite loci of Galba truncatula from different patches in Western Switzerland</td></tr>
<tr><td width="25%"><a href="hierfstat.html">hierfstat</a></td>
<td>General information on the hierfstat package</td></tr>
<tr><td width="25%"><a href="ind.count.html">ind.count</a></td>
<td>individual counts</td></tr>
<tr><td width="25%"><a href="indpca.html">indpca</a></td>
<td>PCA on a matrix of individuals genotypes frequencies</td></tr>
<tr><td width="25%"><a href="mat2vec.html">mat2vec</a></td>
<td>Rewrite a matrix as a vecor</td></tr>
<tr><td width="25%"><a href="nb.alleles.html">nb.alleles</a></td>
<td>Number of different alleles</td></tr>
<tr><td width="25%"><a href="nei.dist.html">nei.dist</a></td>
<td>Estimates Nei's genetic distance</td></tr>
<tr><td width="25%"><a href="pcoa.html">pcoa</a></td>
<td>Principal coordinate analysis</td></tr>
<tr><td width="25%"><a href="indpca.html">plot.indpca</a></td>
<td>PCA on a matrix of individuals genotypes frequencies</td></tr>
<tr><td width="25%"><a href="pop.freq.html">pop.freq</a></td>
<td>Allelic frequencies</td></tr>
<tr><td width="25%"><a href="ppfst.html">pp.fst</a></td>
<td>fst per pair</td></tr>
<tr><td width="25%"><a href="pp.sigma.loc.html">pp.sigma.loc</a></td>
<td>wrapper to return per locus variance components</td></tr>
<tr><td width="25%"><a href="prepdata.html">prepdata</a></td>
<td>Sort and renumber and recode levels for the hierarchical analysis if necessary</td></tr>
<tr><td width="25%"><a href="basic.stats.html">print.bas.stats</a></td>
<td>Basic statistics</td></tr>
<tr><td width="25%"><a href="indpca.html">print.indpca</a></td>
<td>PCA on a matrix of individuals genotypes frequencies</td></tr>
<tr><td width="25%"><a href="print.pp.fst.html">print.pp.fst</a></td>
<td>print function for pp.fst</td></tr>
<tr><td width="25%"><a href="wc.html">print.wc</a></td>
<td>Computes Weir and Cockrham estimates of Fstatistics</td></tr>
<tr><td width="25%"><a href="read.fstat.html">read.fstat</a></td>
<td>Reads data from a FSTAT file</td></tr>
<tr><td width="25%"><a href="read.fstat.data.html">read.fstat.data</a></td>
<td>Reads data from a FSTAT file</td></tr>
<tr><td width="25%"><a href="samp.between.html">samp.between</a></td>
<td>Shuffles a sequence among groups defined by the input vector</td></tr>
<tr><td width="25%"><a href="samp.between.within.html">samp.between.within</a></td>
<td>Shuffles a sequence</td></tr>
<tr><td width="25%"><a href="samp.within.html">samp.within</a></td>
<td>Shuffles a sequence within groups defined by the input vector</td></tr>
<tr><td width="25%"><a href="sim.freq.html">sim.freq</a></td>
<td>Simulates frequencies, for internal use only</td></tr>
<tr><td width="25%"><a href="sim.genot.html">sim.genot</a></td>
<td>Simulates genotypes in an island model at equilibrium</td></tr>
<tr><td width="25%"><a href="subsampind.html">subsampind</a></td>
<td>Subsample a FSTAT data frame</td></tr>
<tr><td width="25%"><a href="test.between.html">test.between</a></td>
<td>Tests the significance of the effect of test.lev on genetic differentiation</td></tr>
<tr><td width="25%"><a href="test.between.within.html">test.between.within</a></td>
<td>Tests the significance of the effect of test.lev on genetic differentiation</td></tr>
<tr><td width="25%"><a href="test.g.html">test.g</a></td>
<td>Tests the significance of the effect of level on genetic differentiation</td></tr>
<tr><td width="25%"><a href="test.within.html">test.within</a></td>
<td>Tests the significance of the effect of inner.level on genetic differentiation within blocks defined by outer.level</td></tr>
<tr><td width="25%"><a href="varcomp.html">varcomp</a></td>
<td>Estimates variance components for each allele of a locus</td></tr>
<tr><td width="25%"><a href="varcomp.glob.html">varcomp.glob</a></td>
<td>Estimate variance components and hierarchical F-statistics over all loci</td></tr>
<tr><td width="25%"><a href="vec2mat.html">vec2mat</a></td>
<td>Reads a vector into a matrix</td></tr>
<tr><td width="25%"><a href="wc.html">wc</a></td>
<td>Computes Weir and Cockrham estimates of Fstatistics</td></tr>
<tr><td width="25%"><a href="writefstat.html">write.fstat</a></td>
<td>Write an Fstat data file</td></tr>
<tr><td width="25%"><a href="write.struct.html">write.struct</a></td>
<td>Write structure file</td></tr>
<tr><td width="25%"><a href="yangex.html">yangex</a></td>
<td>Example data set from Yang (1998) appendix</td></tr>
</table>
</body></html>
